include/communicator_quda.h

```diff
@@ -0,0 +1,939 @@
+#ifdef QUDA_BACKWARDSCPP
+    backward::StackTrace st;
+    st.load_here(32);
+    backward::Printer p;
+    p.print(st, getOutputFile());
+#endif
+    comm_abort_(status);
+  }
+
+#if defined(QMP_COMMS) || defined(MPI_COMMS)
+  MPI_Comm MPI_COMM_HANDLE;
+#endif
+
+#if defined(QMP_COMMS)
+  QMP_comm_t QMP_COMM_HANDLE;
+#endif
+
+  int rank = -1;
+  int size = -1;
+
+  Communicator() { }
+
+  Communicator(Communicator &other, const int *comm_split);
+
+  Communicator(int nDim, const int *commDims, QudaCommsMap rank_from_coords, void *map_data,
+               bool user_set_comm_handle = false, void *user_comm = nullptr);
+
+  ~Communicator() { comm_finalize(); }
+
+  void comm_gather_hostname(char *hostname_recv_buf);
+
+  void comm_gather_gpuid(int *gpuid_recv_buf);
+
+  void comm_init(int ndim, const int *dims, QudaCommsMap rank_from_coords, void *map_data);
+
+  int comm_rank(void);
+
+  int comm_size(void);
+
+  /**
+   * Declare a message handle for sending to a node displaced in (x,y,z,t) according to "displacement"
+   */
+  MsgHandle *comm_declare_send_displaced(void *buffer, const int displacement[], size_t nbytes);
+
+  /**
+   * Declare a message handle for receiving from a node displaced in (x,y,z,t) according to "displacement"
+   */
+  MsgHandle *comm_declare_receive_displaced(void *buffer, const int displacement[], size_t nbytes);
+
+  /**
+   * Declare a message handle for sending to a node displaced in (x,y,z,t) according to "displacement"
+   */
+  MsgHandle *comm_declare_strided_send_displaced(void *buffer, const int displacement[], size_t blksize, int nblocks,
+                                                 size_t stride);
+
+  /**
+   * Declare a message handle for receiving from a node displaced in (x,y,z,t) according to "displacement"
+   */
+  MsgHandle *comm_declare_strided_receive_displaced(void *buffer, const int displacement[], size_t blksize, int nblocks,
+                                                    size_t stride);
+
+  void comm_free(MsgHandle *&mh);
+
+  void comm_start(MsgHandle *mh);
+
+  void comm_wait(MsgHandle *mh);
+
+  int comm_query(MsgHandle *mh);
+
+  template <typename T> T deterministic_reduce(T *array, int n)
+  {
+    std::sort(array, array + n); // sort reduction into ascending order for deterministic reduction
+    return std::accumulate(array, array + n, 0.0);
+  }
+
+  void comm_allreduce(double *data);
+
+  void comm_allreduce_max(double *data);
+
+  void comm_allreduce_min(double *data);
+
+  void comm_allreduce_array(double *data, size_t size);
+
+  void comm_allreduce_max_array(double *data, size_t size);
+
+  void comm_allreduce_int(int *data);
+
+  void comm_allreduce_xor(uint64_t *data);
+
+  /**  broadcast from rank 0 */
+  void comm_broadcast(void *data, size_t nbytes);
+
+  void comm_barrier(void);
+
+  void comm_abort_(int status);
+};
+
+#include <array>
+
+constexpr int nDim = 4;
+using CommKey = std::array<int, nDim>;
+
+void push_to_current(const CommKey &split_key);
```

Comment:
I liked @mathiaswagner 's idea of changing this to  `push_communicator`, I think it'll be clearer